### ISSUE TITLE
relink deletes entire content of yet-to-lazy-load tiddlers!

### ISSUE THREAD

**Issue by avks** [flibbles/tw5-relink#20]
Hello,
I have a fairly large tiddler store — about 1.5k tiddlers. So I have started loading them lazily, using `$/core/save/lazy-all`. I noticed that relink _does_ a very good job in identifying the tiddlers where relinking will required, even if these tiddlers are not lazy loaded... yet. But the problem is if I rename the tiddler, the contents of the associated tiddlers is deleted!

**To repro**

+ Start the tiddlywiki nodejs server with `$/core/save/lazy-all`
+ Create a Tiddler, say `A`, with some text.
+ Create another TIddler, say `B`, with some text and a link to `A`.
+ Now, reload the page and load only `A` (this is required here because `B` won't be loaded if not requested.)
+ Rename `A` to `C`.
   + Relink will notify that the links in `B` will be updated.
+ Now load `B`. See the content gone!

Is there any solution to this? Or am I doing something wrong?

**Comment by flibbles** [comment]
This is definitely a problem that needs addressing, but...

Is there some trick to using lazy-loading I'm not aware of? I'm trying to use it on all of my projects, including tw5.com, but no matter what I do, all non-system tiddlers have blank bodies. They never load.

Is there some setting I'm forgetting? The command I'm using is what tw5.com recommends:
```
tiddlywiki --listen root-tiddler=$:/core/save/lazy-all
```

Nothing works, much less Relink.

Also, I'm not sure it makes sense to use Relink and lazy loading. I've never thought about it before now, but I don't see how Relink could function except by loading _every single tiddler_ so it can examine their bodies, effectively defeating the purpose of lazy loading.

**Comment by avks** [comment]
> loading every single tiddler so it can examine their bodies
But I noticed that relink can correctly identify the tiddlers which needs to be updated (even if those tiddlers are not loaded). How does relink keep track of which tiddlers to update?

**Comment by flibbles** [comment]
I... I don't know. I've never used lazy loading. Relink works by commandeering the relinkTiddler wikimethod.

I'm still trying to replicate the problem. Is the command I posted above the same command you use to run your NodeJS servers? Is there another plugin I need to include?

**Comment by flibbles** [comment]
Okay, I'm making progress now. I'll see what's up.

**Comment by avks** [comment]
> Is the command I posted above the same command you use to run your NodeJS servers?

Yes, exactly that one.

> Is there another plugin I need to include?

I am just using relink and official markdown. Nothing else.

**Comment by flibbles** [comment]
All right. I'm _sort of_ able to replicate this.

When I create A and B, reload, and then rename A to C, B doesn't get changed at all for me. This makes sense, since B is not loaded. Your browser has no way of knowing that it requires update at all.

However, if B also includes A as a tag, such that even the skinny-tiddler requires updating, **now** the text of B gets deleted when I change A's name. Now the skinny tiddler gets updated and saved over the unloaded full-tiddler.

The only logical way to fix this is to make it so **all** tiddlers are loaded when you attempt a rename, because otherwise Relink wouldn't be able to do it's job. It's whole point is to updated all link to the renamed tiddler, not just the ones that have been visited this session.

But then once again this defeats the purpose of using lazy-loading and relink together.

**Comment by flibbles** [comment]
I just tested this _without_ Relink installed. If B has A in its list field, and then you update A to C, the body of B gets destroyed by Tiddlywiki's rudimentory core relinker, the one that only handles `list` and `tags` fields. So this is starting to look more like a design problem with core tiddlywiki.

The best way it would function is if the rename method was actually a server call, so the renaming occurs on the server, but this will involve core work. I can make an issue about this later today.

**Comment by avks** [comment]
Forgive my ignorance but may I suggest you an alternative way?

You said `relink` keeps track of the tiddlers to update by "commandeering the relinkTiddler wikimethod." So would it be possible to load the yet-to-load tiddlers _at the time_ when the name of the tiddler is changed OR _at the time_ when `relink` shows the notification: "The following tiddlers will be updated ..."

**Comment by flibbles** [comment]
That is true. It could be done that way, but since Relink must always look at all tiddlers that might be renamed, which includes all non-system tiddlers. So that means when you rename a tiddler, or you look at the "The following tiddlers..." notification, Tiddlywiki must load **everything**. This effectively defeats the point of lazy-loading.

**Comment by avks** [comment]
>Tiddlywiki must load everything.

Is there anyway to force TW to load **only** those tiddlers that needs to be updated by `relink`?

**Comment by flibbles** [comment]
Only if there's a server task.

Otherwise it has to load the tiddlers in order to examine them and determine if they do need to be updated. This is something tiddlywiki core would need to be capable of.

**Comment by avks** [comment]
> and determine if they do need to be updated. 

But `relink` already does that. `relink` can identify the tiddlers that need to be updated, **even if** they are not loaded.

**Comment by flibbles** [comment]
I actually am not able to replicate that. When I do this:

1. make A
2. make B linking to A *in its body*
3. Reload
4. Open only A and rename it

B does not get changed at all. At least not for me. However, if B links to A in any field apart from the `text` field, then B updates and loses the contents of its body. This is because the not-yet-loaded tiddlers will still have loaded all other fields besides `text`, so Relink is able to check those fields.

**Comment by flibbles** [comment]
[I'm working with core Tiddlywiki](https://github.com/Jermolene/TiddlyWiki5/issues/4767) to come up with the best solution. This might take a while though. So hold tight. In the meantime, I'll maybe make a change soon to Relink so it at least, like, *warns* you about this, or eagerly loads. I dunno. It's super not a good solution though.

**Comment by flibbles** [comment]
I realize by now you've moved on, your kids have grandkids, you're probably enjoying your retirement now, and this issue is but a dream of a dream. But I've resolved it.

...and I've resolved it by saying, "no. I can't do it." After [a talk with Jermolene](https://github.com/Jermolene/TiddlyWiki5/issues/4767), he stated that support for lazy loading will not be elaborated enough to make this fix possible. I have added [documentation](https://flibbles.github.io/tw5-relink/#Lazy%20Loading) explaining that Relink should not be used with lazy loading, and it explains the dangers of doing so.

I know this is not the answer you were hoping for, or even expecting since I'm sure you probably forgot about this thread entirely, but unfortunately it's the one I must give.
